File: Data Structures/Tree/BinarySearchTree.py
# Implementation of binary search tree using python
class Node:
    def __init__(self, data):
        self.data = data
        self.left = None
        self.right = None

class Tree:
    def __init__(self, data=None):
        if data == None:
            self.root = None
        else:
            self.root = Node(data)
    
    def insert(self, data):
        if self.root == None:
            self.root = Node(data)
        else:
            temp = self.root
            while True:
                if data < temp.data:
                    if temp.left == None:
                        temp.left = Node(data)
                        break
                    else:
                        temp = temp.left
                else:
                    if temp.right == None:
                        temp.right = Node(data)
                        break
                    else:
                        temp = temp.right

    def search(self, key):
        if self.root == None:
            print('Tree is empty')
        else:
            temp = self.root
            count = 0
            while True:
                count += 1
                if key == temp.data:
                    return count
                if key < temp.data and temp.left != None:
                    temp = temp.left
                elif key > temp.data and temp.right != None:
                    temp = temp.right
                else:
                    return -1

# driving code
if __name__ == '__main__':
    tree = Tree(30)
    tree.insert(50)
    tree.insert(20)
    tree.insert(70)
    tree.insert(35)
    print(f'35 found at height {tree.search(35)}')
    print(f'20 found at height {tree.search(20)}')
    print(f'40 found at height {tree.search(40)}')

#      30
#     /  \
#    20   50
#        /  \
#       35  70

File: Data Structures/Tree/binary_search_tree.c
#include <stdio.h>
#include <stdlib.h>

// NODE STRUCTURE FOR TREE
struct node{
	int data;
	struct node *left;
	struct node *right;
};

/*	GET A NEW TREE NODE FROM THE HEAP SPACE */
struct node * getnode(int data){
	struct node *new_node = (struct node *) malloc(sizeof(struct node));
	new_node->data	= data;
	new_node->left	= NULL;
	new_node->right	= NULL;
	return new_node;
}

/*	INSERT NEW NODE TO THE BST
	if (head is empty) -> head = newnode()
	elseif (value <= head.data) -> insert(head.left, value)
	else -> insert(head.right, value)
*/
struct node * insert(struct node *root, int value){
	if(root == NULL){
		root = getnode(value);
	}else if(value <= root->data){
		root->left = insert(root->left, value);
	}else{
		root->right = insert(root->right, value);
	}
	return root;
}

/*	SEARCH DATA IN THE BST
	if (tree is empty) -> faliure
	elseif (root == data) -> success
	elseif (data <= root.data) -> recursion(root.left)
	ekse -> recursion(root.right)
*/
int search(struct node *root, int value){
	if(root==NULL){
		return 0;
	}else if(root->data == value){
		return 1;
	}else if(root->data <= value){
		return search(root->left, value);
	}else{
		return search(root->right, value);
	}
	return 0;
}

// DRIVING CODE
void main(){
	struct node *root = NULL;
	root = insert(root, 7);
	root = insert(root, 10);
	root = insert(root, 3);
	root = insert(root, 5);
	printf("%d\n",search(root, 8));
	printf("%d\n",search(root, 7));
}

/*	EXPECTED TREE

      7
     / \
    3   10
     \
      5

*/
